### PR TITLE
use conmon for exec

### PR DESCRIPTION
**Pull request by haircommander**
I haven't gotten to changing CRI-O bits yet, so the tests will definitely fail.
I am putting this out here to get some preliminary feedback on design. There's still some work to do, including:

- [x] fix CRI-O exec to make sure it's backwards compatible
- [x] ~Maybe make exec without sync also use conmon?~
- [x] fix a small tty bug (if I sh in, there's a weird string printed after the prompt)
- [x] generally refactor

### CODE REVIEW

**Comment by haircommander** [comment]
@giuseppe @mrunalp @rhatdan PTAL

**Comment by giuseppe** [comment]
> I haven't gotten to changing CRI-O bits yet, so the tests will definitely fail.

what are the breaking changes in conmon?

**Comment by haircommander** [comment]
Hey @giuseppe @mheon @saschagrunert @mrunalp @rhatdan PTAL at the conmon bits--CI is finally passing here so some feedback would be very appreciated

**Comment by rhatdan** [comment]
Need @mrunalp and @giuseppe and @umohnani8 To review this.  Perhaps @saschagrunert

File: oci/oci_linux.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func sysProcAttrPlatform() *syscall.SysProcAttr {
 
 // newPipe creates a unix socket pair for communication
 func newPipe() (parent, child *os.File, err error) {
-	fds, err := unix.Socketpair(unix.AF_LOCAL, unix.SOCK_STREAM|unix.SOCK_CLOEXEC, 0)
+	fds, err := unix.Socketpair(unix.AF_LOCAL, unix.SOCK_SEQPACKET|unix.SOCK_CLOEXEC, 0)
```
Comment (giuseppe) [inline oci/oci_linux.go]: :+1:

File: conmon/conmon.c [hunk 2]
```diff
@@ -84,10 +85,11 @@ static GOptionEntry opt_entries[] = {
 	{"name", 'n', 0, G_OPTION_ARG_STRING, &opt_name, "Container name", NULL},
 	{"runtime", 'r', 0, G_OPTION_ARG_STRING, &opt_runtime_path, "Runtime path", NULL},
 	{"restore", 0, 0, G_OPTION_ARG_STRING, &opt_restore_path, "Restore a container from a checkpoint", NULL},
-	{"restore-arg", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_restore_args,
```
[the remark this answers is not in the document]
Comment (haircommander) [reply conmon/conmon.c]: done :+1:

File: oci/oci_linux.go (hunk 1 above)
Comment (haircommander) [reply oci/oci_linux.go]: :+1:

File: conmon/conmon.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply conmon/conmon.c]: Did anyone ever use this option?

[the remark this answers is not in the document]
Comment (haircommander) [reply conmon/conmon.c]: podman did, which I changed in the pr over there. Though, there are other things that break backward compatibility other than --restore-arg, so I wasn't inclined to worry about it (most notably: the syncStruct key has changed to "Data" to be less specific.

File: conmon/conmon.c [hunk 3]
```diff
@@ -84,10 +86,13 @@ static GOptionEntry opt_entries[] = {
...
-	 "Additional arg to pass to the restore command. Can be specified multiple times", NULL},
+	{"restore-arg", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_runtime_opts,
+	 "Additional arg to pass to the restore command. Can be specified multiple times. (DEPRECATED)", NULL},
+	{"runtime-opt", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_runtime_opts,
+	 "Additional opts to pass to the restore or exec command. Can be specified multiple times", NULL},
 	{"runtime-arg", 0, 0, G_OPTION_ARG_STRING_ARRAY, &opt_runtime_args,
 	 "Additional arg to pass to the runtime. Can be specified multiple times", NULL},
+	{"attach", 0, 0, G_OPTION_ARG_NONE, &opt_attach, "Attach to an exec session", NULL},
```
Comment (mrunalp) [inline conmon/conmon.c]: nit: we should call this exec-attach

**Comment by rhatdan** [comment]
@haircommander Needs a rebase.

File: conmon/conmon.c (hunk 3 above)
Comment (haircommander) [reply conmon/conmon.c]: fixed

**Pull request closed without merging**